packages/roosterjs-editor-dom/lib/region/getRegionsFromRange.ts

```diff
@@ -0,0 +1,209 @@
+ * @param start A node where full range start from. This may not be the direct node container of range.startContenter.
+ * It is the nearest ancestor which satisfies the InnerSelector of the given region type
+ * @param end A node where full range end from. This may not be the direct node container of range.endContenter.
+ * It is the nearest ancestor which satisfies the InnerSelector of the given region type
+ * @param started Whether we have already hit the start node
+ */
+function iterateNodes(
+    creator: (rootNode: HTMLElement, nodeBefore?: Node, nodeAfter?: Node) => Region,
+    boundary: Boundary,
+    start: Node,
+    end: Node,
+    started?: boolean
+): [Region[], boolean, boolean] {
+    started = started || boundary.innerNode == start;
+    let ended = false;
+    const { children, innerNode } = boundary;
+    let regions: Region[] = [];
+
+    if (children.length == 0) {
+        regions.push(creator(innerNode));
+    } else {
+        // Need to run one more time to add region after all children
+        for (let i = 0; i <= children.length && !ended; i++) {
+            const { outerNode, boundaries } = children[i] || {};
+            const previousOuterNode = children[i - 1]?.outerNode;
+            if (started) {
+                regions.push(creator(innerNode, previousOuterNode, outerNode));
+            }
+
+            boundaries?.forEach(child => {
+                let newRegions: Region[];
+                [newRegions, started, ended] = iterateNodes(creator, child, start, end, started);
+                regions = regions.concat(newRegions);
+            });
+        }
+    }
+
+    return [regions, started, ended || innerNode == end];
+}
+
+/**
+ * Check if the given nodes combination is valid to create a region.
+ * A combination is valid when:
+ * 1. Root node is not null and is not empty. And
+ * 2. For nodeBefore and nodeAfter, each of them should be either null or contained by root node. And
+ * 3. If none of nodeBefore and nodeAfter is null, the should not contain each other, and there should be
+ * node between them.
+ * @param root Root node of region
+ * @param nodeBefore The boundary node before the region under root
+ * @param nodeAfter The boundary node after the region under root
+ * @param skipTags Tags to skip
+ */
+function areNodesValid(root: Node, nodeBefore: Node, nodeAfter: Node, skipTags: string[]) {
+    if (!root || !root.firstChild) {
+        return false;
+    } else {
+        const firstNodeOfRegion = nodeBefore && getNextLeafSibling(root, nodeBefore, skipTags);
+        const lastNodeOfRegion = nodeAfter && getPreviousLeafSibling(root, nodeAfter, skipTags);
+        const firstNodeValid =
+            !nodeBefore || (contains(root, nodeBefore) && contains(root, firstNodeOfRegion));
+        const lastNodeValid =
+            !nodeAfter || (contains(root, nodeAfter) && contains(root, lastNodeOfRegion));
+        const bothValid =
+            !nodeBefore ||
```

Comment:
Does the bothValid check need to re-check if `nodeBefore` and `nodeAfter` are falsy? Aren't those covered by the previous two checks already, and since all the values are `&&`, you already know whether or not those values are or are not falsy?

Reply:
Without checking falsy for nodeBefore and nodeAfter, if one of them is null, contains() will return false, then bothValid will be false, so the final result is also false.

Reply:
But if one of them is null, then `!null` will be true, and the whole boolean check will short circuit and `bothValid` will be true, at least as written.